### PR TITLE
[ESQL] Remove unused NullEquals class

### PR DESCRIPTION
**Pull request by not-napoleon**
Although the ES|QL language doesn't support a null equals operator, we had a placeholder class for it with no `@Evaluator` methods defined. I presume this was to match the QL `BinaryComparisonOperator` enum, although I wasn't involved in the initial discussions around that choice.  This PR removes the unused code, and associated references to it.

### CODE REVIEW

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRulesTests.java [hunk 1]
```diff
@@ -227,17 +222,6 @@ public void testDualEqualsConjunction() {
         assertEquals(FALSE, exp);
     }
 
-    // a <=> 1 AND a <=> 2 -> FALSE
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRulesTests.java:230]: These test cases were just copied over from the QL version of the optimizer rules.

**Review by alex-spies — APPROVED** [review]
LGTM, thanks @not-napoleon !

**Comment by not-napoleon** [comment]
The backport conflicts are related to https://github.com/elastic/elasticsearch/pull/105407#pullrequestreview-2031250852 which wasn't backported.  I don't think this is actually all that important to backport, so I'm just going to leave it as 8.15 only.

**Review by astefan — APPROVED** [review]
LGTM

To give a bit of background on what NullEquals is:
- it was introduced in ES SQL as part of https://github.com/elastic/elasticsearch/issues/35871
- it returns `true` / `false` on equality against `null` values. Because SQL has a 3 value logic, any equality with `null` would result in `null`. This operator helps having a truly null equality

In ESQL we have the `IS NULL` / `IS NOT NULL` predicates that do the same thing. ES SQL also has these two predicates, but offered a shortcut in the form of an operator (NullEquals `<=>`).

ESQL also allows `field == null` BUT it's requiring a specific casting ie `field == null::integer` and this equality returns `null`. In a sense, `field == null` it's the ESQL variant of ES SQL's `field = null`.

**Pull request merged**